### ISSUE TITLE
3-Divide & Bind Your Attention for Improved Generative Semantic Nursing

### ISSUE THREAD

**Issue by mjmaher987** [sharif-ml-lab/visual_compositional_generation_survey#3]
## Task: Read the Paper and Do the following 4 steps
1. Highlight the paper systematically (**green** for important notes - **blue** for very important notes that worth mentioning in the final survey - **red** for the sentences that we don’t understand or need further attention in the future) [In Zotero].
2. Write the Wikipedia summary of the paper in the GitHub issue (_comment in this issue_).
3. Extract the information in the [excel file](https://docs.google.com/spreadsheets/d/15_sh6TqtUvZIJo1-RZxQtsG0Nvg0vPG_IFKCWiGMjdY/edit?usp=sharing).
4. Add the related sentences to the [overleaf file](https://www.overleaf.com/8167246262ccxtrzztjvpz). Note that when you add the sentence, don’t forget to write the reference in the format of ”[ #reference number that is mentioned in the Google Doc - Name of the Author (who added this to the overleaf) ]” at the end of the sentence.

**Comment by saba-200979** [comment]
### Introduction

There are two well-known semantic issues in text-to-image synthesis, i.e., “missing objects” and “attribute binding”. “Missing objects” refers to the phenomenon that not all objects mentioned in the input text faithfully appear in the image. “Attribute binding” represents the critical compositionality problem that the attribute information, e.g., color or texture, is not properly aligned to the corresponding object or is wrongly attached to the other object. To mitigate these issues, recent work Attend & Excite (A&E)  has introduced the concept of Generative Semantic Nursing (GSN). The core idea lies in updating latent codes on-the-fly such that the semantic information in the given text can be better incorporated within pretrained synthesis models. To enforce the object occurrence, A&E defined a loss objective that attempts to maximize the maximum attention value for each object token. In this work, it proposes a novel objective function for GSN. it maximizes the total variation of the attention map to prompt multiple, spatially distinct attention excitations. By spatially distributing the attention for each token, it enables the generation of all objects mentioned in the prompt, even under high token competition. Intuitively, this corresponds to dividing the attention map into multiple regions. Besides, to mitigate the attribute binding issue, it proposes a Jensen-Shannon divergence (JSD) based binding loss to explicitly align the distribution between the excitation of each object and its attributes. Thus, this paper terms its method Divide & Bind.

### Contributions

Main contributions can be summarized as:
(i) We propose a novel total-variation-based attendance loss enabling the presence of multiple objects in the generated image. 
(ii) We propose a JSD-based attribute binding loss for faithful attribute binding. 
(iii) the approach exhibits outstanding capability of generating images fully adhering to the prompt, outperforming A&E on several benchmarks involving complex descriptions.

### Preliminaries

**Stable Diffusion (SD):**
We implement our method based on the open-source state-of-the-art T2I model SD Rombach et al. (2022), which belongs to the family of latent diffusion models (LDMs). LDMs are two-stage methods, consisting of an autoencoder and a diffusion model trained in the latent space. In the first stage, the encoder E transforms the given image x into a latent code z = E(x), then z is mapped back to the image space by the decoder D. The autoencoder is trained to reconstruct the given image, i.e. D(E(x)) ≈ x. In the second stage, a diffusion model is trained in the latent space of the autoencoder. During training, we gradually add noise to the original latent $z_0$ with time, resulting in $z_t$. Then the UNet denoiser $ϵ_θ$ is trained with a denoising objective to predict the noise ϵ that is added to $z_0$:

$L = E[∥ϵ − ϵ_θ(zt, c, t)∥^2]$

where c is the conditional information.

**Cross-Attention in Stable Diffusion:**
In SD, a frozen CLIP text encoder is adopted to embed the text prompt P into a sequential embedding as condition c, which is then injected into UNet through cross-attention (CA) to synthesize text-complied images. The CA layers take the encoded text embedding and project it into queries Q and values V. The keys K are mapped from the intermediate features of UNet. The attention maps $A_t$ can be reshaped into $R^{h×w×L}$, where h,w is the resolution of the feature map, and L is the sequence length of the text embedding. Further, we denote the cross-attention map that corresponds to the $s_{th}$ text token as $A^s_t \in  R^{h×w}$.

### Method

![image1](C:\Users\Asus\Desktop\img1.PNG)  

Given the recognized significance of the cross-attention maps in guiding semantic synthesis, our method aims at optimizing the latent code at inference time to excite them based on the text tokens. We employ the generative semantic nursing (GSN) method for latent code optimization and propose a novel loss formulation. It consists of two parts, i.e. divide and bind, which
encourages object occurrence and attribute binding respectively.

**GENERATIVE SEMANTIC NURSING (GSN)**
To improve the semantic guidance in SD during inference, one pragmatic way is via latent code
optimization at each time step of sampling, 

$z'_t← z_t − α_t · ∇L$

where $α_t$ is the updating rate and L is the loss to encourage the faithfulness between the image and
text description, e.g. object attendances and attribute binding. GSN has the advantage of avoiding fine-tuning SD.

**DIVIDE & BIND**
Our proposed method Divide & Bind consists of a novel objective for GSN:

$min L = min L_{attend} + \lambda L_{bind}$

which has two parts, the attendance loss $L_{attend}$ and the binding loss $L_{bind}$ that respectively enforce
the object attendance and attribute binding. λ is the weighting factor.

**Divide for Attendance** 

The attendance loss $L_{attend}$ is to incentivize the presence of the objects,
thus is applied to the text tokens associated with objects S,

$L_{attend} = −min TV (A^s_t), TV (A^s_t) =\sum |A^s_t[i + 1, j] − A^s_t[i, j]| + |A^s_t[i, j + 1] − A^s_t[i, j]|$

where $A^s_t[i, j]$ denotes the attention value of the $s-{th}$ token at the specific location [i, j] and time step t. The loss formulation is based on the finite differences approximation of the total variation (TV)  along the spatial dimensions. It is evaluated for each object token and we take the smallest value, i.e., representing the worst case among all object tokens. Taking the negative TV as the loss, we essentially maximize the TV for latent optimization. Since TV is essentially computed as a form of summation across the spatial dimension, it encourages large activation differences across many neighboring at different spatial locations rather than a single one, thus not only having one high activation region but also many of them. Such an activation pattern in the space resembles to dividing it into different regions. The model can select some of them to display the object with single or even multiple attendances. This way, conflicts between different objects that compete for the same region can be more easily resolved.


**Attribute Binding Regularization.**
 
In addition to the object attendance, the given attribute information, e.g. color or material, should be appropriately attached to the corresponding object. We denote the attention map of the object token and its attribute token as $A^s_t$ and $A^r_t$, respectively. For attribute binding, it is desirable that $A^r_t$ and $A^s_t$ are spatially well-aligned, i.e. high activation regions
of both tokens are largely overlapped. To this end, we introduce L_{bind}. To explicitly encourage such alignment, we can then minimize the symmetric similarity measure Jensen–Shannon divergence (JSD) between these two distributions:

$L_{bind} = JSD(A^r_t∥A^s_t)$

GPT-3 can automatically extract the desired nouns and adjectives for new input prompts.

### Experiments

**Benchmarks** 
We conduct an exhaustive evaluation of seven prompt sets as summarized in the following table.

![image2](C:\Users\Asus\Desktop\img2.PNG)  

**Evaluation metrics**

To quantitatively evaluate the performance of our method, we used the text-text similarity from Chefer et al. (2023) and the recently introduced TIFA score.

![image3](C:\Users\Asus\Desktop\img3.PNG)  

**Overall algorithm** 
![image4](C:\Users\Asus\Desktop\img4.PNG)  

### Limitations

Despite improved semantic guidance, it is yet difficult to generate extremely rare or implausible cases, e.g., unusual color binding “a gray apple”. Our method may generate such objects together with the common one, e.g., generating a green apple and a gray apple in the same image. As we use the pretrained model without fine-tuning, some data bias is inevitably inherited.
Another issue is miscounting: more instances may be generated than they should. We attribute the miscounting to the imprecise language understanding limited by the CLIP text encoder.

### Comparison

![image5](C:\Users\Asus\Desktop\img5.PNG)
